dashboard/tests/test_model_layer.py

```diff
@@ -25,4 +35,162 @@ def test_get_sessions_country_stats(self):
+        self._create_session(
+            'session key 2', 'node key', 'service type', now, 3, 4
+        )
+        self._create_session(
+            'session key 3', 'node key', 'other service type', now, -100, -100
+        )
+        self._create_session(
+            'session key 4', 'other node key', 'service type', now, -100, -100
+        )
+        _, total_bytes = aggregate_node_data_from_sessions(
+            'node key',
+            'service type',
+            now,
+            now + second
+        )
+        self.assertEqual(10, total_bytes)
+
+    def test_where_dates_mismatch(self):
+        self._create_session(
+            'session key 1', 'node key', 'service type', now - second, 1, 2
+        )
+        self._create_session(
+            'session key 2', 'node key', 'service type', now + second, 1, 2
+        )
+        _, total_bytes = aggregate_node_data_from_sessions(
+            'node key',
+            'service type',
+            now,
+            now + second
+        )
+        self.assertEqual(0, total_bytes)
+
+    @staticmethod
+    def _create_session(session_key, node_key, service_type,
+                        updated_at, bytes_sent, bytes_received):
+        session = models.Session(session_key)
+        session.node_key = node_key
+        session.service_type = service_type
+        session.client_updated_at = updated_at
+        session.client_bytes_sent = bytes_sent
+        session.client_bytes_received = bytes_received
+        get_db().session.add(session)
+
+
+class TestAggregateSessionsCount(TestCase):
+    def test_match_query_params(self):
+        self._create_session(
+            'session key 1', 'node key', 'service type', now
+        )
+        self._create_session(
+            'session key 2', 'other node key', 'service type', now
+        )
+        self._create_session(
+            'session key 3', 'node key', 'other service type', now
+        )
+        self._create_session(
+            'session key 4', 'node key', 'service type', now - second
+        )
+        self._create_session(
+            'session key 5', 'node key', 'service type', now + second
+        )
+        self._create_session(
+            'session key 6', 'node key', 'service type', now
+        )
+        sessions_count, _ = aggregate_node_data_from_sessions(
+            'node key',
+            'service type',
+            now,
+            now + second
+        )
+        self.assertEqual(2, sessions_count)
+
+    @staticmethod
+    def _create_session(session_key, node_key, service_type, updated_at):
+        session = models.Session(session_key)
+        session.node_key = node_key
+        session.service_type = service_type
+        session.client_updated_at = updated_at
+        get_db().session.add(session)
+
+
+class TestGetNodeHoursOnline(TestCase):
+    def test_returns_one_hour(self):
+        for x in range(31):
+            self._create_node_availability('node key', 'service type', now)
```

Comment:
We are assuming that node sends availability messages once a minute - if node modifies this behaviour, it's very easy to get infinite availability hours, i.e. by sending 100 availability messages on startup.

Is this really safe to assume? Maybe we can somehow easily protect from this? (I wasn't able to think of easy solution myself, so it's a question :D )

Reply:
this should probably be limited by the API - aka do not accept an availability ping more often than a predefined time interval.